egs2/TEMPLATE/asr1/pyscripts/utils/dialog_eval/LLM_Metrics.py

```diff
@@ -0,0 +1,124 @@
+        metric_arr.append(metric)
+
+        str1 += f"{metric_name}: {metric}\n"
+    str1 += f"VERT: {round(gmean(metric_arr), 2)}\n"
+    return str1
+
+
+def bert_score(total_response_arr):
+    def cosine_similarity_context_response(context, response, model, tokenizer):
+        # Tokenize and encode both context and response
+        context_inputs = tokenizer(context, return_tensors="pt", truncation=True)
+        response_inputs = tokenizer(response, return_tensors="pt", truncation=True)
+        for k in context_inputs:
+            context_inputs[k] = context_inputs[k].cuda()
+        for k in response_inputs:
+            response_inputs[k] = response_inputs[k].cuda()
+
+        # Get embeddings from the model
+        with torch.no_grad():
+            context_embedding = model(**context_inputs).last_hidden_state.mean(dim=1)
+            response_embedding = model(**response_inputs).last_hidden_state.mean(dim=1)
+
+        # Compute cosine similarity
+        similarity = cosine_similarity(
+            context_embedding.cpu().numpy(), response_embedding.cpu().numpy()
+        )
+        return similarity[0][0]
+
+    bert_model_name = "bert-base-uncased"
+    bert_model = AutoModel.from_pretrained(bert_model_name).cuda()
+    bert_tokenizer = AutoTokenizer.from_pretrained(bert_model_name)
+    similarity = cosine_similarity_context_response(
+        " ".join(total_response_arr[:-1]),
+        total_response_arr[-1],
+        bert_model,
+        bert_tokenizer,
+    )
+    return f"Cosine Similarity: {similarity*100:.2f}" + "\n"
+
+
+def DialoGPT_perplexity(user_utterance, response):
+    def evaluate_response_with_dialoGPT(context, response, model, tokenizer):
+        """
+        Evaluate the appropriateness of a response based on the
+        given context using DialoGPT.
+
+        Args:
+            context (str): The dialogue context (previous conversation).
+            response (str): The generated response to evaluate.
+            model: Pre-trained DialoGPT model.
+            tokenizer: Corresponding tokenizer for the DialoGPT model.
+
+        Returns:
+            float: Perplexity score of the response given the context.
+        """
+        model.eval()
+
+        # Combine context and response as input
+        input_text = context + tokenizer.eos_token + response + tokenizer.eos_token
+        inputs = tokenizer(input_text, return_tensors="pt", truncation=True)
+        inputs["input_ids"] = inputs["input_ids"].cuda()
+        inputs["attention_mask"] = inputs["attention_mask"].cuda()
+        # import pdb;pdb.set_trace()
+
+        # Compute model outputs and loss
+        with torch.no_grad():
+            outputs = model(**inputs, labels=inputs["input_ids"].cuda())
+            loss = outputs.loss
+
+        # Calculate perplexity
+        perplexity = torch.exp(loss)
+        return perplexity.cpu().item()
+
+    # Load DialoGPT model and tokenizer
+    model_name = (
```

Comment:
Turn it on a argument with default vaule

Reply:
Sounds good! Added!